### PR TITLE
Draft: feat: governance proposals init plus indexer

### PR DESCRIPTION
**Pull request by NamelsKing**

### CODE REVIEW

File: serverless.yml [hunk 1]
```diff
@@ -441,6 +443,35 @@ resources:
           - AttributeName: 'id'
             AttributeType: 'S'
 
+    governanceProposalsData:
+      Type: AWS::DynamoDB::Table
+      Properties:
+        TableName: ${self:custom.table.GOVERNANCE_PROPOSALS_DATA}
+        BillingMode: PAY_PER_REQUEST
+        KeySchema:
```
Comment (mrbasado) [inline serverless.yml:451]: did we not want network as part of the key?

File: src/aws/models/governance-proposals.model.ts [hunk 2]
```diff
@@ -0,0 +1,70 @@
+import { embed } from '@aws/dynamodb-data-mapper';
+import { attribute, hashKey, rangeKey, table } from '@aws/dynamodb-data-mapper-annotations';
+import { Network } from '@badger-dao/sdk';
+
+import { GOVERNANCE_PROPOSALS_DATA } from '../../config/constants';
+import { GovernanceProposalsDisputes } from './governance-proposals-disputes.interface';
+import { GovernanceProposalsStatuses } from './governance-proposals-statuses.interface';
+
+@table(GOVERNANCE_PROPOSALS_DATA)
+export class GovernanceProposals {
```
Comment (mrbasado) [inline src/aws/models/governance-proposals.model.ts:10]: am i missing something here. does this account for proposals with more than 1 command? + ScheduleBatch()

File: src/indexers/governance-proposals-indexer.ts [hunk 3]
```diff
@@ -0,0 +1,251 @@
...
+import { getSupportedChains } from '../chains/chains.utils';
+import { getLastProposalUpdateBlock } from '../governance/governance.utils';
+
+export async function updateGovernanceProposals() {
+  console.info('Updating governance proposals started');
+
+  const mapper = getDataMapper();
+
+  for (const chain of getSupportedChains()) {
+    const timelockAddress = chain.sdk.governance.timelockAddress;
```
Comment (mrbasado) [inline src/indexers/governance-proposals-indexer.ts:22]: this 1:1 mapping of chain to timelock addr makes having a test addr not available

File: serverless.yml (hunk 1 above)
Comment (NamelsKing) [reply serverless.yml:451]: we have it in `GlobalSecondaryIndexes`, u can have only 1 primary index in ddb

File: src/indexers/governance-proposals-indexer.ts (hunk 3 above)
Comment (NamelsKing) [reply src/indexers/governance-proposals-indexer.ts:22]: this is till we will add timelocks to the registry and sdk

File: src/aws/models/governance-proposals.model.ts (hunk 2 above)
Comment (NamelsKing) [reply src/aws/models/governance-proposals.model.ts:10]: hmmm, idk, havent unpacked calldata yet, but as i can see, its `bytes`, not `bytes[]`
```
    /**
     * @dev Emitted when a call is scheduled as part of operation `id`.
     */
    event CallScheduled(
        bytes32 indexed id,
        uint256 indexed index,
        address target,
        uint256 value,
        bytes data,
        bytes32 predecessor,
        uint256 readyTime,
        address sender,
        string status
    );
```

Comment (NamelsKing) [reply src/aws/models/governance-proposals.model.ts:10]: About ScheduleBatch ye sure, it emits the same event as regular schedule, so should do

Comment (mrbasado) [reply src/aws/models/governance-proposals.model.ts:10]: ah right, if they are part of a batch their id will all be the same, with the index showing the order

File: serverless.yml (hunk 1 above)
Comment (mrbasado) [reply serverless.yml:451]: oh i'm blind

**Review by mrbasado — DISMISSED** [review]

File: src/indexers/governance-proposals-indexer.ts [hunk 4]
```diff
@@ -0,0 +1,329 @@
...
+        updatedProposalsStatuses.push(...statuses);
+        updatedProposalsDisputes.push(...disputes);
+        continue;
+      }
+
+      await saveChildProposal(chain, timelockAddress, proposal, updatedProposals, mapper);
+    }
+
+    try {
+      updatedProposals.length > 0 && (await mapper.put(updatedProposals));
```
Comment (axejintao) [inline src/indexers/governance-proposals-indexer.ts:89]: should this be batch put?

File: src/indexers/governance-proposals-indexer.ts [hunk 5]
```diff
@@ -0,0 +1,329 @@
...
+    const disputesToUpdate = proposalsDisputed.filter((e) => !updatedProposalsDisputes.includes(e));
+
+    if (statusesToUpdate.length === 0 && disputesToUpdate.length === 0) {
+      console.info(`No new statuses or disputes for ${chain.network}`);
+      continue;
+    }
+
+    for (const status of statusesToUpdate) {
+      let proposal!: GovernanceProposals;
+      const proposalDdbIdx = GovernanceProposals.genIdx(chain.network, timelockAddress, status.args.id);
```
Comment (axejintao) [inline src/indexers/governance-proposals-indexer.ts:105]: we could probably chunk this into a helper function

File: src/indexers/governance-proposals-indexer.ts [hunk 6]
```diff
@@ -0,0 +1,329 @@
...
+      } catch (err) {
+        console.error({ message: `Unable to save governance proposal ${proposalDdbIdx}`, err });
+      }
+
+      console.info(`New status ${status.transactionHash} for proposal ${proposalDdbIdx} added`);
+    }
+
+    for (const dispute of disputesToUpdate) {
+      let proposal!: GovernanceProposals;
+      const proposalDdbIdx = GovernanceProposals.genIdx(chain.network, timelockAddress, dispute.args.id);
```
Comment (axejintao) [inline src/indexers/governance-proposals-indexer.ts:149]: we could probably chunk this into a helper function

**Review by axejintao — COMMENTED** [review]
soem comments

File: src/indexers/governance-proposals-indexer.ts (hunk 3 above)
Comment (NamelsKing) [reply src/indexers/governance-proposals-indexer.ts:26]: i believe its about mapping on sdk, and mrbasado wants to save possibility of having test addr

File: src/aws/models/governance-proposals.model.ts [hunk 7]
```diff
@@ -0,0 +1,74 @@
...
+
+  @attribute()
+  creationBlock!: number;
+
+  @attribute({
+    indexKeyConfigurations: {
+      IndexGovernanceProposalsUpdateBlock: 'RANGE',
+    },
+  })
+  updateBlock!: number;
```
[the remark this answers is not in the document]
Comment (NamelsKing) [reply src/aws/models/governance-proposals.model.ts:60]: `idx` is main hash key and its composite

File: src/aws/models/governance-proposals.model.ts [hunk 8]
```diff
@@ -0,0 +1,74 @@
...
+
+import { GOVERNANCE_PROPOSALS_DATA } from '../../config/constants';
+import { GovernanceProposalsChild } from './governance-proposals-child.interface';
+import { GovernanceProposalsDisputes } from './governance-proposals-disputes.interface';
+import { GovernanceProposalsStatuses } from './governance-proposals-statuses.interface';
+
+@table(GOVERNANCE_PROPOSALS_DATA)
+export class GovernanceProposals {
+  @hashKey()
+  idx!: string;
```
[the remark this answers is not in the document]
Comment (NamelsKing) [reply src/aws/models/governance-proposals.model.ts:13]: https://github.com/Badger-Finance/badger-api/pull/1591/files#diff-afd4c3a626e3564100f31d5c78cd159044876f3a031a6746f9f1ffc6d7ce4455R71-R73

its composite key ser

File: src/governance/governance.utils.ts [hunk 9]
```diff
@@ -0,0 +1,22 @@
+import { Network } from '@badger-dao/sdk';
+
+import { getDataMapper } from '../aws/dynamodb.utils';
+import { GovernanceProposals } from '../aws/models/governance-proposals.model';
+
+export async function getLastProposalUpdateBlock(network: Network) {
```
[the remark this answers is not in the document]
Comment (NamelsKing) [reply src/governance/governance.utils.ts:6]: ill do all test as follow ups, like i said, now trying to deliver endpoint, so we can start make something on UI. Will cover everything, no worries

File: src/indexers/governance-proposals-indexer.ts [hunk 10]
```diff
@@ -0,0 +1,329 @@
...
+    const timelockAddress = chain.sdk.governance.timelockAddress;
+
+    if (!timelockAddress) {
+      console.info(`No timelock address for ${chain.network}`);
+      continue;
+    }
+
+    const lastScannedBlock = await getLastProposalUpdateBlock(chain.network);
+
+    const governanceProxy = new GovernanceProxyMock(chain.sdk);
```
[the remark this answers is not in the document]
Comment (NamelsKing) [reply src/indexers/governance-proposals-indexer.ts:35]: its a helper class from main package, i'll think how to do this in more appropriate way

File: src/indexers/governance-proposals-indexer.ts [hunk 11]
```diff
@@ -0,0 +1,329 @@
...
+      }
+
+      console.info(`New dispute ${dispute.transactionHash} for proposal ${proposalDdbIdx} added`);
+    }
+  }
+
+  console.info('Updating governance proposals ended');
+}
+
+async function saveRelativeProposal(
```
[the remark this answers is not in the document]
Comment (NamelsKing) [reply src/indexers/governance-proposals-indexer.ts:196]: k sure

File: src/indexers/governance-proposals-indexer.ts [hunk 12]
```diff
@@ -0,0 +1,329 @@
...
+          chain,
+          timelockAddress,
+          proposal,
+          proposalsStatusesChanged,
+          proposalsDisputed,
+        );
+        updatedProposals.push(proposalModel);
+        updatedProposalsStatuses.push(...statuses);
+        updatedProposalsDisputes.push(...disputes);
+        continue;
```
[the remark this answers is not in the document]
Comment (NamelsKing) [reply src/indexers/governance-proposals-indexer.ts:82]: ye, i kinda like to return/continue vs rising cyclomatic complexity... but can do

File: src/indexers/governance-proposals-indexer.ts (hunk 4 above)
Comment (NamelsKing) [reply src/indexers/governance-proposals-indexer.ts:89]: i wonder this API all the time, ye you are right, thanks

File: src/indexers/governance-proposals-indexer.ts (hunk 5 above)
Comment (NamelsKing) [reply src/indexers/governance-proposals-indexer.ts:105]: yes, make sense to me

File: src/aws/models/governance-proposals.model.ts (hunk 8 above)
[the remark this answers is not in the document]
Comment (axejintao) [reply src/aws/models/governance-proposals.model.ts:13]: missed that thanks

**Review by axejintao — APPROVED** [review]
comments are nits, will merge this one to allow for fixes + tests follow ups

File: src/governance/responses/proposal.list.response.ts [hunk 13]
```diff
@@ -0,0 +1,35 @@
+import { Description, Example, Property, Title } from '@tsed/schema';
+
+import { GovernanceProposals } from '../../aws/models/governance-proposals.model';
+import { GovernanceProposalsList } from '../interfaces/governance.proposals.list';
+
+export class ProposalListResponse implements GovernanceProposalsList {
```
[the remark this answers is not in the document]
Comment (NamelsKing) [reply src/governance/responses/proposal.list.response.ts:6]: i take 1st one

**Pull request merged**